contracts/utils/helpers.cairo

```diff
@@ -0,0 +1,386 @@
+    user_id : felt,
+    generation : felt,
+    cell_index : felt,
+    state : felt
+):
+end
+
+##########
+
+func pay{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt, 
+        credit_requirement : felt
+    ):
+    ERC20._burn(user, Uint256(credit_requirement, 0))
+    return ()
+end
+
+func reward_user{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt
+    ):
+    ERC20._mint(user, Uint256(1, 0))
+    return ()
+end
+
+func ensure_user{
+        syscall_ptr : felt*
+    }(
+    ) -> (
+        caller : felt
+    ):
+    let (caller) = get_caller_address()
+    with_attr error_message(
+        "User not authenticated"
+    ):
+        assert_not_zero(caller)
+    end
+    return (caller)
+end
+
+func assert_valid_new_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game : felt
+    ):
+    alloc_locals
+    let (local game_info) = stored_game.read(
+        game_id=game, generation=0)
+    let game_id = game_info[1]
+
+    with_attr error_message(
+        "Game with genesis state {game} already exists"
+    ):
+        assert_not_equal(game_id, game)
+    end
+
+    let (high, _) = split_felt(game)
+    let (invalid_bits, _) = unsigned_div_rem(high, 2**97)
+    with_attr error_message(
+        "Game {game} is not a valid game"
+    ):
+        assert invalid_bits = 0
+    end
+    return ()
+end
+
+func get_last_state{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }() -> (
+        current_generation : felt,
+        current_state : felt
+    ):
+    let (generation) = current_generation.read(
+        game_id=INFINITE_GAME_GENESIS
+    )
+    let (game_info) = stored_game.read(
+        game_id=INFINITE_GAME_GENESIS,
+        generation=generation
+    )
+    return(generation, game_info[1])
+end
+
+func assert_valid_cell_index{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        cell_index : felt
+    ):
+    with_attr error_message(
+        "Cell index {cell_index} out of range"
+    ):
+        assert_le_felt(cell_index, DIM*DIM-1)
+    end
+
+    return ()
+end
+
+func create_new_game{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_state : felt,
+        user_id : felt
+    ):
+    save_game(
+        game_id=game_state,
+        generation=0,
+        user_id=user_id,
+        packed_game=game_state
+    )
+    save_generation(
+        game_id=game_state,
+        generation=0
+    )
+    game_created.emit(
+        owner_id=user_id,
+        game_id=game_state,
+        state=game_state
+    )
+    return ()
+end
+
+# User input may override state to make a cell alive.
+func activate_cell{
```

Comment:
I think splitting functions for common logic, creator mode and infinite mode would be great. Currently when reading you switch context depending on type of game.

Reply:
Ok, so I have sorted functions by their purpose and added comments in the code to mark where each section starts; I think that should be enough, but I you'd rather move them to separate files let me know :+1:

Reply:
Ok, that's probably enough.